include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h

```diff
@@ -538,25 +494,42 @@ struct __get_sycl_range
     auto
     operator()(_Iter __first, _Iter __last)
         -> ::std::enable_if_t<is_temp_buff<_Iter>::value && !is_zip<_Iter>::value && !is_permutation<_Iter>::value,
-                              __buffer_holder<val_t<_Iter>, AccMode>>
+                              __range_holder<oneapi::dpl::__ranges::all_view<val_t<_Iter>, AccMode>>>
     {
         static_assert(!oneapi::dpl::__internal::is_const_iterator<_Iter>::value || AccMode == sycl::access::mode::read,
                       "Should be non-const iterator for a modifying algorithm.");
 
         assert(__first < __last);
 
-        using copy_direct_tag = ::std::integral_constant<bool, AccMode == sycl::access::mode::read_write ||
-                                                                   AccMode == sycl::access::mode::read>;
-        using copy_back_tag = ::std::integral_constant<bool, AccMode == sycl::access::mode::read_write ||
-                                                                 AccMode == sycl::access::mode::write>;
-
-        auto buf = copy_direct(__first, __last, copy_direct_tag());
-        buf = copy_back(__first, buf, copy_back_tag());
-
-        using BufferType = oneapi::dpl::__internal::__lifetime_keeper<decltype(buf)>;
-        m_buffers.push_back(::std::make_shared<BufferType>(buf));
-
-        return __buffer_holder<val_t<_Iter>, AccMode>{buf};
+        using _T = val_t<_Iter>;
+
+        constexpr bool __is_copy_direct =
+            AccMode == sycl::access::mode::read_write || AccMode == sycl::access::mode::read;
+        constexpr bool __is_copy_back =
+            AccMode == sycl::access::mode::read_write || AccMode == sycl::access::mode::write;
+
+        auto __get_buf = [&]()
+        {
+            if constexpr(__is_copy_direct)
+                //buffer will wait and copying on destructor
+                return sycl::buffer<_T, 1>(std::addressof(*__first), __last - __first);
```

Comment:
We have avoided `std::addressof(*__first)` for input iterators in other places with the justification that we don't know really what the `operator*()` will return.  It could be a proxy object rvalue for instance.  
This overload is supposed to be for iterators to host data (or iterators to `std::vector`).  However, this is only detected by process of elimination. 

We discussed this last week and you mentioned that the initialization of `sycl::buffers` with iterators rather than a pointer to host memory will cause an extra host-to-host copy to occur, which can be avoided when initializing a buffer with a pointer to host memory.  

Are there implementations of `std::vector` for which this would not work?  I don't think we get any guarantees that the data range `host_vec.begin()` -> `host_vec.end()` is a contiguous block of memory.  However, in practice I think this is going to be the case. 

What about `host_vec.rbegin()`?  I think we "support" this case, but it seems reverse iterators to host vectors would be broken after this change.

While I don't want to incur the extra host-to-host copy, I think we may have to for correctness. We would need to rely on `set_final_data` and `set_write_back` as well if we use iterators to initialize the buffer.

Reply:
> I don't think we get any guarantees that the data range `host_vec.begin()` -> `host_vec.end()` is a contiguous block of memory

Actually, for `std::vector `container we have such guarantee:
https://en.cppreference.com/w/cpp/container/vector
`The elements are stored contiguously, which means that elements can be accessed not only through iterators, but also using offsets to regular pointers to elements. This means that a pointer to an element of a vector may be passed to any function that expects a pointer to an element of an array.`

Reply:
> So, in my opinion, an ideal solution here - if `std::addressof(*__first)` is valid, we use exclusive mem access buffer `sycl::buffer(ptr,N)`, if is not - we do fallback to `sycl::buffer(It1, it2)` usage.

I agree, I think that would be ideal.  Is it possible to properly determine if `std::addressof` is valid?

Reply:
> Is it possible to properly determine if `std::addressof` is valid?

yes, it is possible via SFINAE approach.
Additionally, we have to check "contiguous" property and to have custom code for reverse iterator

Reply:
Ok, that makes sense.  Since `std::addressof` deletes the rvalue call, that takes care of those, then if we check contiguity and specialize for reverse iterator, I think that will be sufficient. 

Since we don't have c++20, but we do know that `std::addressof` is valid, for contiguity we can just check confirm that the addresses are the same distance away as the iterators, and perhaps an incremented iterator's address is 1 away from itself or something like that.

Reply:
> an incremented iterator's address is 1 away from itself or something like that.

I thought about such check... but it doesn't guarantees contiguity in general..

Reply:
You are correct, but without c++20 concepts I'm not seeing a robust to confirm this.  If both the full range is the appropriate distance and a single increment is then I would feel pretty good about it.  Of course, you could design an iterator to thwart this, but it does not seem likely (and perhaps we don't care about such cases because they would not be officially supported by our specification / documentation).